### PR TITLE
Make the iBridges search more easy to use

### LINKED ISSUE #229: Improve search_data results
Proposal: Change the search results to return an `IrodsPath` or `CachedIrodsPath` instead of a dictionary. This should contain all the information that the user will need, right?

### LINKED ISSUE #230: Add search functionality to the CLI
Probably after #229 is done.

### LINKED ISSUE #233: Refactor Search
Input: collection in which to search (IrodsPath), name: collection or data name (str), metadata (dict), checksum (str)
output: List of IrodsPath

### PR DESCRIPTION
**Pull request by qubixes**
Significant API change for the search function.

TODO:

- [x] Documentation
- [x] Tutorials
- [x] CLI
- [x] Update/improve tests

Fixes #233
Fixes #229
Fixes #230

### CODE REVIEW

File: ibridges/search.py [hunk 1]
```diff
@@ -47,90 +60,146 @@ def search_data(
...
+        if isinstance(units, str):
+            _units_filter(units, *queries)
+        else:
+            for sub_units in units:
+                _units_filter(sub_units, *queries)
 
     results = []
-    if checksum:
+    if checksum is not None:
+        if not checksum.startswith("sha2:"):
```
Comment (chStaiger) [inline ibridges/search.py:148]: I understand that you want to free the user of prefixing the checksum with sha. However iRODS servers know two checksums, sha2 and md5. Only the sha2 is prefixed with sha2 while md5 is the real md5 checksum without any additional prefix.

So in iRODS instances where md5 checksums are configured the search will fail to retrieve any results. :(

File: ibridges/search.py [hunk 2]
```diff
@@ -47,90 +60,146 @@ def search_data(
...
+    # Filter only paths according to the search path.
+    coll_query = coll_query.filter(icat.LIKE(icat.COLL_NAME, f"{path}/%"))
+    data_query = data_query.filter(icat.LIKE(icat.COLL_NAME, f"{path}/%"))
+    data_name_query.filter(icat.LIKE(icat.COLL_NAME, f"{path}"))
+    queries = [coll_query, data_query, data_name_query]
+
+    if path_pattern is not None:
+        _path_filter(path, path_pattern, *queries)
+
+    if key is not ...:
```
Comment (chStaiger) [inline ibridges/search.py:125]: Nice!!! I am beginning to like ellipsises!

File: ibridges/search.py [hunk 3]
```diff
@@ -47,90 +60,146 @@ def search_data(
...
+        return pattern
+    return f"%/{pattern}"
+
+def _path_filter(root_path, path_pattern, coll_query, data_query, data_name_query):
+    coll_query.filter(icat.LIKE(icat.COLL_NAME, _prefix_wildcard(path_pattern)))
+    split_pat = path_pattern.rsplit("/", maxsplit=1)
+    data_query.filter(icat.LIKE(icat.DATA_NAME, split_pat[-1]))
+    data_name_query.filter(icat.LIKE(icat.DATA_NAME, split_pat[-1]))
+
+    if len(split_pat) == 2:
```
Comment (chStaiger) [inline ibridges/search.py:189]: I think this will return more results than the search defined?

If I am looking for `/my_coll/%/books`
then `_prefix_wildcard(split_pat[0])` will not only return  all `books` collections under `my_coll`, but everything?! Do I understand the code correctly?

Comment (chStaiger) [reply ibridges/search.py:189]: Or is there then another filter on the query that specifies the data name or the coll name somewhere up in the code?

File: ibridges/search.py [hunk 4]
```diff
@@ -47,90 +60,146 @@ def search_data(
...
+        data_query.filter(icat.LIKE(icat.COLL_NAME, _prefix_wildcard(split_pat[0])))
+        data_name_query.filter(icat.LIKE(icat.COLL_NAME, _prefix_wildcard(split_pat[0])))
+
+
+def _key_filter(key, *queries):
```
Comment (chStaiger) [inline ibridges/search.py:194]: These three functions are very useful!!! Thanks for restructuring my Gordian knot!

**Review by chStaiger — CHANGES_REQUESTED** [review]
I like the new structure of the code. It makes it much more readable.
I think we need to add some good test examples to see whether the search retrieves all or not too many results.

- path_pattern = %/my_coll/%books, where books is a collection
- path_pattern = %/my_coll/%book.txt
- Testing the `%` in the metadata and in the checksum.
- ...

File: ibridges/search.py (hunk 3 above)
Comment (qubixes) [reply ibridges/search.py:189]: For your query I would use `search_data(session, IrodsPath(session, "/my_coll"), path_pattern="books")`, so you wouldn't need any wildcards for that anymore. Otherwise `search_data(session, IrodsPath(session, "/"), path_pattern="/my_coll/%/books")` should also work I think.

Comment (qubixes) [reply ibridges/search.py:189]: But indeed good to test these edge cases.

**Review by chStaiger — COMMENTED** [review]
On the whole I must say, the code and the usage of the function is really nice. 
I just find the creation of the MetaSearch object less intuitiv than creating a dictionary and passing that to the search function to pass the metadata criteria.

**Comment by qubixes** [comment]
> On the whole I must say, the code and the usage of the function is really nice. I just find the creation of the MetaSearch object less intuitiv than creating a dictionary and passing that to the search function to pass the metadata criteria.

You can also provide a tuple, however for the validation you are on your own then.

**Comment by chStaiger** [comment]
> > On the whole I must say, the code and the usage of the function is really nice. I just find the creation of the MetaSearch object less intuitiv than creating a dictionary and passing that to the search function to pass the metadata criteria.
> 
> You can also provide a tuple, however for the validation you are on your own then.

Can we add the validation and transformation into a MetaSearch object to the search function if a list of tuples for metadata is provided? I think that would make it easier for our users who are not so familiar with programming but still write little python scripts.

**Comment by qubixes** [comment]
> > > On the whole I must say, the code and the usage of the function is really nice. I just find the creation of the MetaSearch object less intuitiv than creating a dictionary and passing that to the search function to pass the metadata criteria.
> > 
> > 
> > You can also provide a tuple, however for the validation you are on your own then.
> 
> Can we add the validation and transformation into a MetaSearch object to the search function if a list of tuples for metadata is provided? I think that would make it easier for our users who are not so familiar with programming but still write little python scripts.

I feel like those users will simply copy paste whatever is put in the tutorial. I don't understand how it is more difficult though. Compare:

```python
from ibridges import MetaSearch

search(path, metadata=MetaSearch(value="some_value")
```
or
```python
search(path, metadata=[(..., "some_value", ...)])
```
Yes, it's shorter, but it looks much more like magic to me.

**Review by chStaiger — APPROVED** [review]
Some small changes in wording, small bugfixes in the tutorials. After that is fixed: Approved.

Nice work!!!

File: docker/irods_client/tests/test_search.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    (MetaSearch(value="10"), True),
+    (MetaSearch(key="Random"), False),
+    (MetaSearch(key="Author", value="10"), False),
+    (MetaSearch(key="Author", units="kg"), False),
+    (MetaSearch(key="Mass", units=None), False),
+    ([MetaSearch(key="Author"), MetaSearch(value="10")], True),
+    ([MetaSearch("Author"), MetaSearch("Random")], False),
+])
+@mark.parametrize("item_name", ["collection", "dataobject"])
+def test_find_meta(session, item_name, request, metadata, is_found):
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docker/irods_client/tests/test_search.py:52]: Inside the test function, see a few lines below it.

File: ibridges/search.py [hunk 6]
```diff
@@ -2,19 +2,48 @@
...
+
+
+def search_data(  # pylint: disable=too-many-branches
     session: Session,
     path: Optional[Union[str, IrodsPath]] = None,
+    path_pattern: Optional[str] = None,
     checksum: Optional[str] = None,
-    key_vals: Optional[dict] = None,
-) -> list[dict]:
+    metadata: Union[None, MetaSearch, list[MetaSearch], list[tuple]] = None,
```
[the remark this answers is not in the document]
Comment (qubixes) [reply ibridges/search.py:44]: MetaSearch is a (named) tuple, so actually list[tuple] will work regardless. I think we should keep it like this, but use MetaSearch in all our documentation.

File: docs/source/irods_search.rst [hunk 7]
```diff
@@ -40,18 +32,17 @@ We need to create a python dictionary which contains the metadata keys ad their
 
 .. code-block:: python
 
-	key_vals = {'key': 'value'}
-	search_data(session, key_vals = key_vals)
+	search_data(session, metadata=MetaSearch(key="key", value="value"))
 	key_vals = {'key': ''}
-	search_data(session, key_vals = key_vals)
+	search_data(session, metadata=MetaSearch(key="key", value=""))
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/irods_search.rst]: Done.

File: docs/source/irods_search.rst [hunk 8]
```diff
@@ -40,18 +32,17 @@ We need to create a python dictionary which contains the metadata keys ad their
 
 .. code-block:: python
 
-	key_vals = {'key': 'value'}
-	search_data(session, key_vals = key_vals)
+	search_data(session, metadata=MetaSearch(key="key", value="value"))
 	key_vals = {'key': ''}
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/irods_search.rst]: Done.

File: docs/source/irods_search.rst [hunk 9]
```diff
@@ -40,18 +32,17 @@ We need to create a python dictionary which contains the metadata keys ad their
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/irods_search.rst:42]: Added.

File: docs/source/irods_search.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/irods_search.rst]: Done.

**Comment by jjkoehorst** [comment]
> I like the new structure of the code. It makes it much more readable. I think we need to add some good test examples to see whether the search retrieves all or not too many results.
> 
> * path_pattern = %/my_coll/%books, where books is a collection
> * path_pattern = %/my_coll/%book.txt
> * Testing the `%` in the metadata and in the checksum.
> * ...

Will this also work?

path_pattern = %/my_coll/%books, where books is a data object? 

In response to #243. For me personally it took me a while to figure out why can't it find %fancy%path%bla.txt? as I was missing a magic / ? Would it be possible to have a collection= and data= optional search so you can more clearly specify what you want or is that making things more difficult for users?

**Comment by chStaiger** [comment]
> > I like the new structure of the code. It makes it much more readable. I think we need to add some good test examples to see whether the search retrieves all or not too many results.
> > 
> > * path_pattern = %/my_coll/%books, where books is a collection
> > * path_pattern = %/my_coll/%book.txt
> > * Testing the `%` in the metadata and in the checksum.
> > * ...
> 
> Will this also work?
> 
> path_pattern = %/my_coll/%books, where books is a data object?
> 
> In response to #243. For me personally it took me a while to figure out why can't it find %fancy%path%bla.txt? as I was missing a magic / ? Would it be possible to have a collection= and data= optional search so you can more clearly specify what you want or is that making things more difficult for users?

Yes that is true. To denote that you are looking for subcollections, you will need to add the "/". I also do not particularly like it, but it is in line with the icommands.
We will see what we can do about it, as the wildcard `%` does not seem to include the `/`.

**Comment by jjkoehorst** [comment]
How is it inline with iCommands? Just curious as iquest:

`iquest --no-page "SELECT COLL_NAME, DATA_NAME WHERE COLL_NAME LIKE '%humann3%' AND DATA_NAME LIKE '%'"` makes the distinction?

**Comment by chStaiger** [comment]
> How is it inline with iCommands? Just curious as iquest:
> 
> `iquest --no-page "SELECT COLL_NAME, DATA_NAME WHERE COLL_NAME LIKE '%humann3%' AND DATA_NAME LIKE '%'"` makes the distinction?

I just did some testing. You do not need to the "/". You will get all collections, subcollection and data objects with the substring denoted in `path_pattern`. Here the output of our CLI:

```
christine@sibelius iBridges % ibridges search --path-pattern "%demo%"
/nluu12p/home/research-test-christine/demo/demofile.txt
/nluu12p/home/research-test-christine/new_coll/demofile.txt
/nluu12p/home/research-test-christine/demo
/nluu12p/home/research-test-christine/demofile.txt
```

**Comment by trel** [comment]
Does it also pick up this file?
```
/nluu12p/home/research-test-christine/demo/another.txt
```

**Comment by trel** [comment]
and this collection?
```
/nluu12p/home/research-test-christine/demon/
```

**Comment by chStaiger** [comment]
@trel  The first one not, but the demon yes. 
The problem we are thinking about is, that the pattern can be split out over the COLL_NAME and the DATA_NAME.

**Comment by trel** [comment]
I think you may have to perform two queries?   One for collections and one for data objects?  @d-w-moore?

**Comment by chStaiger** [comment]
I already identified three:
```
"Get all data objects on a path -pattern"
"select COLL_NAME, DATA_NAME where COLL_NAME like '%demo%'"
"Get all data objects adhering to the path-pattern"
iquest "select COLL_NAME, DATA_NAME where DATA_NAME like '%demo%'"
"Get all collections adhering to the path-pattern"
iquest "select COLL_NAME, DATA_NAME where COLL_NAME like '%demo%'"
```

But if we assume that the path-pattern can be distributed over collection paths and object names, we end up with a lot of queries ;)

**Comment by qubixes** [comment]
> I think you may have to perform two queries? One for collections and one for data objects? @d-w-moore?

I think more depending on the number of wildcards in the pattern. The problem is splitting up the pattern into one part for the collection and another part for the data object. Let's say we have the pattern "%a%b%c":

- %a%b%c collection
- %a%b% collection, c data object
- %a%b collection, %c data object
- etc

I think we will have to require users to put a slash in the pattern to tell us where the separation is between the collection part of the pattern and where the data object part is of the pattern. We could allow multiple patterns, but I'm not sure how much that would actually solve in this case. Ah @chStaiger is still awake as well..

**Comment by trel** [comment]
Is it limiting to disallow forward slashes in the search?

Can you just have them insert alphanumeric characters and then *you* prepend and append the % before searching?  And then you're back to just two queries...

Scenario... looking for counterexamples...

User enters `foo` and you search for:
1) `select COLL_NAME, DATA_NAME where DATA_NAME like '%foo%'`
2) `select COLL_NAME where COLL_NAME like '%foo%'`

Is that sufficient?  Until you give them an 'advanced search' where they can actually specify more clearly what they are looking for with their own column names and wildcards?

**Comment by chStaiger** [comment]
That does not cover all scenarios. Assume someone is looking only for '.txt' files in a collection. Those two searches would not be able to pick that up. So we would at least need the third query where we compare the DATA_NAME to the pattern.

**Comment by qubixes** [comment]
@chStaiger I would prefer not to have more parameters, especially if it isn't immediately clear to our (novice) users what it means. 

I think we should keep it the way it is for now (the behavior hasn't changed from the current release version). All queries that a user might want to do (and are possible) are currently covered as far as I can see. I do think it will be useful to think about how this can be documented. Perhaps we can have a page with wildcard examples showing what will match and what will not match.
